Tests/WorkspaceTests/ToolsVersionSpecificationRewriterTests.swift

```diff
@@ -17,185 +17,170 @@
-        }
+        let inMemoryFileSystem = InMemoryFileSystem()
 
-        // Contents with invalid tools version specification (ignoring the validity of the version specifier).
-        rewriteToolsVersionSpecificationToDefaultManifest(
-            content: """
-            // swift-tool-version:3.1.2
-            ...
-            """
-        ) { result in
-            XCTAssertEqual(result, "// swift-tools-version:4.1.2\n// swift-tool-version:3.1.2\n...")
-        }
+        let manifestFilePath = AbsolutePath("/pkg/Package.swift")
 
-        // Contents with invalid version specifier.
-        rewriteToolsVersionSpecificationToDefaultManifest(
-            content: """
-            // swift-tools-version:3.1.2
-            ...
-            """
-        ) { result in
-            XCTAssertEqual(result, "// swift-tools-version:4.1.2\n...")
-        }
+        try inMemoryFileSystem.createDirectory(manifestFilePath.parentDirectory, recursive: true)
+        try inMemoryFileSystem.writeFileContents(manifestFilePath, string: content)
 
-        // Contents with invalid version specifier and some meta data.
-        rewriteToolsVersionSpecificationToDefaultManifest(
-            content: """
-            // swift-tools-version:3.1.2
-            ...
-            """
-        ) { result in
-            // Note: Right now we lose the metadata but if we ever start using it, we should preserve it.
-            XCTAssertEqual(result, "// swift-tools-version:4.1.2\n...")
-        }
+        try ToolsVersionSpecificationWriter.rewriteSpecification(
+            manifestDirectory: manifestFilePath.parentDirectory,
+            toolsVersion: version,
+            fileSystem: inMemoryFileSystem
+        )
 
-        // Try to write a version with prerelease and build meta data.
-        let toolsVersion = ToolsVersion(version: "4.1.2-alpha.beta+sha.1234")
-        rewriteToolsVersionSpecificationToDefaultManifest(
-            content: "let package = ... \n",
-            version: toolsVersion
-        ) { result in
-            XCTAssertEqual(result, "// swift-tools-version:4.1.2\nlet package = ... \n")
-        }
+        // resultHandler(try inMemoryFileSystem.readFileContents(manifestFilePath))
+        let actual = try #require(try inMemoryFileSystem.readFileContents(manifestFilePath))
+        #expect(actual.validDescription == expected, "Actual is not expected")
     }
-    
-    func testManifestAccessFailures() throws {
+
+    @Test
+    func manifestAccessFailures() throws {
         let toolsVersion = ToolsVersion.v5_3
-        
+
         let inMemoryFileSystem = InMemoryFileSystem()
         let manifestFilePath = AbsolutePath("/pkg/Package.swift/Package.swift")
         try inMemoryFileSystem.createDirectory(manifestFilePath.parentDirectory, recursive: true) // /pkg/Package.swift/
-        
+
         // Test `ManifestAccessError.Kind.isADirectory`
-        XCTAssertThrowsError(
+
+        #expect{
```

Comment:
As above

Reply:
I changed this to 
```
        #expect(throws: ToolsVersionSpecificationWriter.ManifestAccessError(.isADirectory, at: manifestFilePath.parentDirectory)) {
            try ToolsVersionSpecificationWriter.rewriteSpecification(
                manifestDirectory: manifestFilePath.parentDirectory.parentDirectory, // /pkg/
                toolsVersion: toolsVersion,
                fileSystem: inMemoryFileSystem
            )
        }
```
and got the following compilation error

```
/Users/bkhouri/Documents/git/public/swiftlang/swift-package-manager/Tests/WorkspaceTests/ToolsVersionSpecificationRewriterTests.swift:153:9: error: macro 'expect(throws:_:sourceLocation:performing:)' requires that 'ToolsVersionSpecificationWriter.ManifestAccessError' conform to 'Equatable'
151 |             return isExpectedKind && isExpectedDescription
152 |         }
153 |         #expect(throws: ToolsVersionSpecificationWriter.ManifestAccessError(.isADirectory, at: manifestFilePath.parentDirectory)) {
    |         `- error: macro 'expect(throws:_:sourceLocation:performing:)' requires that 'ToolsVersionSpecificationWriter.ManifestAccessError' conform to 'Equatable'
154 |             try ToolsVersionSpecificationWriter.rewriteSpecification(
155 |                 manifestDirectory: manifestFilePath.parentDirectory.parentDirectory, // /pkg/

Testing.expect:1:40: note: where 'E' = 'ToolsVersionSpecificationWriter.ManifestAccessError'
1 | @freestanding(expression) public macro expect<E, R>(throws error: E, _ comment: @autoclosure () -> Testing.Comment? = nil, sourceLocation: Testing.SourceLocation = #_sourceLocation, performing expression: () async throws -> R) = #externalMacro(module: "TestingMacros", type: "ExpectMacro") where E : Equatable, E : Error
  |                                        `- note: where 'E' = 'ToolsVersionSpecificationWriter.ManifestAccessError'
[5/6] Compiling WorkspaceTests ToolsVersionSpecificationRewriterTests.swift
```

Am I doing something wrong here?  If feels odd modifying the production code just to make the test compile. is this the norm in Swift?

Reply:
It means this `Error` type doesn't conform to `Equatable`, so we can't just compare instances of it. You have three choices:

1. Modify the error type to be `Equatable` (may not be possible depending on the cases),
2. Use `#expect(throws: ToolsVersionSpecificationWriter.ManifestAccessError.self)` so it only filters on type rather than exact error (which is often sufficient for test purposes), or
3. Continue using `#expect {} throws: {}`, except I'm [deprecating](https://github.com/swiftlang/swift-testing/blob/jgrynspan/return-errors-from-expect-throws/Documentation/Proposals/NNNN-return-errors-from-expect-throws.md) it. :)